[UserEvents] Add end-to-end runtime test

mdh1418 opened the pull request:
With user_events support added in https://github.com/dotnet/runtime/pull/115265, this PR looks to test a basic end-to-end user_events scenario.

## Alternative testing approaches considered

### **Existing EventPipe runtime tests**
Existing EventPipe tests under `src/tests/tracing/eventpipe` are incompatible with testing the user_events scenario due to:

1. Starting EventPipeSessions through DiagnosticClient ❌ 
DiagnosticClient does not have the support to send the IPC command to start a user_events based EventPipe session, because it requires the user_events_data file descriptor to be sent using SCM_RIGHTS (see https://github.com/dotnet/diagnostics/blob/main/documentation/design-docs/ipc-protocol.md#passing_file_descriptor).

2. Using an EventPipeEventSource to validate events streamed through EventPipe ❌ 
User_events based EventPipe sessions do not stream events. Instead, events are written to configured TraceFS tracepoints, and currently only RecordTrace from https://github.com/microsoft/one-collect/ is capable of generating `.nettrace` traces from tracepoint user_events.

### **Native EventPipe Unit Tests**
There are Mono Native EventPipe tests under `src/mono/mono/eventpipe/test` that are not hooked up to CI. These unit tests are built through linking the shared EventPipe interface library against [Mono's EventPipe runtime shims](https://github.com/dotnet/runtime/tree/main/src/mono/mono/eventpipe) and using [Mono's test runner](https://github.com/dotnet/runtime/tree/main/src/mono/mono/eglib/test). To update these unit tests into the [standard runtime tests structure](https://github.com/dotnet/runtime/tree/main/src/tests), a **larger investment** is needed to either migrate EventPipe from using runtime shims to a OS Pal source shared by coreclr/nativeaot/mono (see https://github.com/dotnet/runtime/pull/118874#discussion_r2474264532) or build an EventPipe shared library specifically for the runtime test using a runtime-agnostic shim.
As existing mono unit tests don't currently test IPC commands, coupled with no existing runtime infrastructure to read events from tracepoints, there would be even more work on top of updating mono native eventpipe unit tests to even test the user_events scenario.

## End-to-End Testing Added
A low-cost approach to testing .NET Runtime's user_events functionality leverages RecordTrace from https://github.com/microsoft/one-collect/, which is already capable of starting user_events based EventPipe sessions and generating `.nettrace`s. (Note: [dotnet-trace wraps around RecordTrace](https://github.com/dotnet/diagnostics/pull/5570))
Despite adding an external dependency which allows RecordTrace failures to fail the end-to-end test, user_events was initially added with the intent to depend on RecordTrace for the end-to-end scenario, and there are no other ways to functionally test a user_events based eventpipe session.

### Approach
1. Start Tracee app
2. Start tracing with RecordTrace + [dotnet-common profile](https://github.com/dotnet/diagnostics/blob/main/src/Tools/dotnet-trace/CommandLine/Commands/ListProfilesCommandHandler.cs) script 
3. Stop RecordTrace (triggers .nettrace generation) and Tracee app
4. Validate the `.nettrace` for particular events from Tracee app

### Dependencies:
- CI runs the runtime test in an environment that supports user_events
- CI runs the runtime test with permissions to access `user_events_data`.
- Microsoft.OneCollect.RecordTrace (transitively resolved through a dotnet diagnostics public feed)
- Microsoft.Diagnostics.Tracing.TraceEvent 3.1.24+ (to read [NetTrace V6](https://github.com/microsoft/perfview/blob/main/src/TraceEvent/EventPipe/NetTraceFormat.md))

mdh1418 committed:
[UserEvents] Add end-to-end runtime test

mdh1418 committed:
Merge remote-tracking branch 'upstream/main' into user_events_functional_runtime_test

mdh1418 committed:
Cleanup test props

mdh1418 committed:
Add ProcessStartInfo and using keyword

mdh1418 committed:
Fix process exit detection

mdh1418 committed:
Retain nettrace test asset

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+
+            if (!File.Exists(recordTracePath) || !File.Exists(scriptFilePath))
+            {
+                Console.WriteLine("record-trace or dotnet-common.script not found. Test cannot run.");
```
Its helpful to print the exact path we didn't find. (In general its helpful for the test logging to err towards being overly verbose to make failure diagnosis easier)

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+
+            using Process traceeProcess = Process.Start(traceeStartInfo);
+            using Process recordTraceProcess = Process.Start(recordTraceStartInfo);
+            recordTraceProcess.OutputDataReceived += (_, args) => Console.WriteLine($"[record-trace] {args.Data}");
```
We should redirect the tracee output as well to ensure we aren't losing useful error diagnostics that it might print.

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+                return -1;
+            }
+
+            ProcessStartInfo traceeStartInfo = new();
```
I'd suggest logging the command-line for both processes being launched.

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+            recordTraceProcess.ErrorDataReceived += (_, args) => Console.Error.WriteLine($"[record-trace] {args.Data}");
+            recordTraceProcess.BeginErrorReadLine();
+
+            if (!traceeProcess.HasExited && !traceeProcess.WaitForExit(15000))
```
I'd suggest logging that the test is waiting for the tracee to exit, and certainly log if the wait times out and we kill it.

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+        private const int SIGINT = 2;
+
+        [DllImport("libc", SetLastError = true)]
+        private static extern int kill(int pid, int sig);
```
```suggestion
        private static extern int Kill(int pid, int sig);
```

mdh1418 replied:
Done

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+        }
+
+        public static int TestEntryPoint()
+        {
```
We should add checks for:
1. process is elevated
2. OS is Linux
3. user_events are supported

Its likely at some point this test will be run in the wrong environment and the logs should make it trivial to diagnose.

mdh1418 replied:
There is now a UserEventsRequirements helper that checks if the environment meets the requirements.

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+
+        private static bool ValidateTraceeEvents(string traceFilePath)
+        {
+            string etlxPath = TraceLog.CreateFromEventPipeDataFile(traceFilePath);
```
You can parse the .nettrace file directly using EventPipeEventSource in TraceEvent. This avoids creating a 2nd file that the test also needs to clean up.

mdh1418 replied:
For some reason when I tried it last, it didn't work (pilot error), switched to using EventPipeEventSource. Right now the events are "unknown" with id. Maybe its cause I'm using the Dynamic parser? I'll look into TraceEvent more closely

mdh1418 replied:
it turned out it was just that AllocationSampled event is too new that PerfView doesn't know about it, Using an EventPipeEventSource by default now.

noahfalk commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+            recordTraceStartInfo.RedirectStandardError = true;
+
+            using Process traceeProcess = Process.Start(traceeStartInfo);
+            using Process recordTraceProcess = Process.Start(recordTraceStartInfo);
```
To ensure tracer observes the tracee we should start the tracer process first.

lateralusX commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+            bool startEventFound = false;
+            bool stopEventFound = false;
+
+            source.AllEvents += (TraceEvent e) =>
```
Is the plan to add more tests here that checks for metadata/fields, rundown events, callstacks etc or should we add some basic verification to this test or plan to extend existing EventPipe tests to also work over UserEvents?

mdh1418 replied:
This was mainly to add a basic verification that the end to end runtime side (from accepting the ipc message to writing to the tracepoints worked). Since User_events is built on EventPipe, my initial thought is that duplicating the existing eventpipe tests for user events wouldn't be adding anything. I think we can add more tests later on, but not sure what coverage is good and reasonable. I'm not even sure yet if our CI machines have user_events, or if they run with elevated privileges, so this was mainly to see if we can have a basic E2E test going.

lateralusX replied:
We should definitely not duplicate but maybe look on extending the existing event pipe tests to run over user events + additional validate logic but agree that is something we could look at later.

So, if this is mainly a smoke test then maybe we should make sure we at least hit things we know is handled in the one-collect library, during the work we hit a number of things that needed special attention, like activity id's, custom metadata and potential stack traces. Right now, these only tests one runtime start/stop event fired under very unique circumstances. Maybe we should do some short multi-threading scenario as well, making sure we won't hit any races in the code path unique to user events?

mdh1418 committed:
Address feedback

mdh1418 committed:
Fix RecordTrace package resolving

mdh1418 commented on src/tests/tracing/eventpipe/userevents/userevents.cs:
```diff
@@ -0,0 +1,120 @@
+            }
+
+            // Until record-trace supports duration, the only way to stop it is to send SIGINT (ctrl+c)
+            kill(recordTraceProcess.Id, SIGINT);
```
Added, thanks!

mdh1418 commented on src/tests/tracing/eventpipe/userevents/usereventstracee.cs:
```diff
@@ -0,0 +1,31 @@
+        public static void Run()
+        {
+            long startTimestamp = Stopwatch.GetTimestamp();
+            long targetTicks = Stopwatch.Frequency * 10; // 10s
```
Changed

mdh1418 commented on src/tests/tracing/eventpipe/userevents/userevents.csproj:
```diff
@@ -0,0 +1,36 @@
+    <ReferenceXUnitWrapperGenerator>false</ReferenceXUnitWrapperGenerator>
+    <TargetFrameworkIdentifier>.NETCoreApp</TargetFrameworkIdentifier>
+    <IlasmRoundTripIncompatible>true</IlasmRoundTripIncompatible>
+    <MicrosoftOneCollectRecordTraceVersion>0.1.32221</MicrosoftOneCollectRecordTraceVersion>
```
Changed to using a test local NuGet.Config to add dotnet-diagnostics-tests as a source.

mdh1418 committed:
Alternative user_events_data access check

mdh1418 committed:
Fix record-trace startup

mdh1418 committed:
Detect record-trace unexpected exit and flush output

mdh1418 committed:
Fix record-trace output

mdh1418 committed:
Reduce enabled keywords

For a basic E2E test, just enabling GC is sufficient.

Locally running this test, occasionally the test failed
because it didn't have the expected events. It's not
clear why, but with a brief tracee timeout of 1s,
collecting dotnet-trace's default keywords and enabling
all associated events is overkill for a fast E2E test.

mdh1418 committed:
Switch to AllocationSampled Event

Even after switching to just enabling GCKeyword,
the 1s tracee app had a 1% failure rate locally.
On the other hand, AllocationSampled had no
failures after 1000 runs.

mdh1418 committed:
Attempt to copy userevents assets for CI

mdh1418 committed:
Enable Helix tests to run record-trace

mdh1418 committed:
Add requirements check and upload trace

mdh1418 committed:
[TEMP] disable all other pipelines

mdh1418 commented:
Looks like the reason the .NET runtime events aren't being captured in the .nettrace is because a session isn't actually being started.
On helix machines, the diagnostic port is created under helix's provisioned environment's tempdirectory which is of the form `/datadisks/disk1/work/<workID>/t/`. [RecordTrace currently only scans /tmp/](https://github.com/microsoft/one-collect/issues/183) for these diagnostic ports. I'm planning on adding a config value for eventpipe/userevents debugging for more stresslogs for better diagnostics on whether the point of failure is in the runtime side or external.

mdh1418 committed:
Configure DiagnosticPort creation for RecordTrace

mdh1418 committed:
[Test] Extend UserEvents to reusable TestRunner

mdh1418 committed:
[Test] Remove eventpipe embedded userevents tests

mdh1418 committed:
Merge remote-tracking branch 'upstream/main' into user_events_functional_runtime_test

mdh1418 committed:
Include userevents_common in helix payload

mdh1418 commented:
Closing in favor of https://github.com/dotnet/runtime/pull/122134, which has a cleaner commit history and less noisy (for now) discussion thread (from all of the helix test commits)
